File: sge/src/sge/core/window.cpp
/*
   Copyright 2021 Nora Beda

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

#include "sgepch.h"
#include "sge/core/window.h"
#include "sge/renderer/renderer.h"
#ifdef SGE_PLATFORM_DESKTOP
#include "sge/platform/desktop/desktop_window.h"
#endif
namespace sge {
    std::unique_ptr<window> window::create(const std::string& title, uint32_t width, uint32_t height) {
        window* instance = nullptr;

#ifdef SGE_PLATFORM_DESKTOP
        instance = new desktop_window(title, width, height);
#endif

        if (instance == nullptr) {
            throw std::runtime_error("no implemented platform was selected!");
        }
        return std::unique_ptr<window>(instance);
    }

    void window::create_swapchain() {
        this->m_swapchain = std::unique_ptr<swapchain>(renderer::create_swapchain(*this));
    }
};

File: csharp/SGE.Scriptcore/Window.cs
/*
   Copyright 2022 Nora Beda and SGE contributors

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

using System;
using System.Collections.Generic;

namespace SGE
{
    public enum DialogMode
    {
        Open = 0,
        Save
    }

    public sealed class DialogFilter
    {
        public DialogFilter(string filter)
        {
            Name = Filter = filter;
        }

        public DialogFilter(string name, string filter)
        {
            Name = name;
            Filter = filter;
        }

        public string Name { get; }
        public string Filter { get; }
    }

    public sealed class Window
    {
        public Window(string title, uint width, uint height)
        {
            CoreInternalCalls.CreateWindow(title, width, height, out mAddress);
            if (mAddress == IntPtr.Zero)
            {
                throw new ArgumentException("Could not create window!");
            }
        }

        internal Window(IntPtr address)
        {
            mAddress = address;
            CoreInternalCalls.AddRef_window(mAddress);
        }

        ~Window() => CoreInternalCalls.RemoveRef_window(mAddress);

        public uint Width => CoreInternalCalls.GetWindowWidth(mAddress);
        public uint Height => CoreInternalCalls.GetWindowHeight(mAddress);

        public string FileDialog(DialogMode mode, IEnumerable<DialogFilter> filters)
        {
            var filterList = new List<DialogFilter>(filters);
            return CoreInternalCalls.WindowFileDialog(mAddress, mode, filterList);
        }

        private readonly IntPtr mAddress;
    }
}
